### PR TITLE
EXL-341 - Implemented PrinterCounters

### PR DESCRIPTION
**Pull request by zsso**
- Implemented PrinterCounters
- Created PrinterDailyAverageCalculator cron job

### CODE REVIEW

File: plugin/inc/installsteps/crontasksinstallstep.class.php [hunk 1]
```diff
@@ -43,10 +43,21 @@ class CronTasksInstallStep
...
+        ], [
+            'itemtype' => 'PluginIservicePrinterDailyAverageCalculator',
+            'name' => 'PrinterDailyAverageCalculator',
+            'mode' => CronTask::MODE_EXTERNAL,
+            'frequency' => 86400,
+            'param' => 0,
+            'state' => 1,
+            'hourmin' => 0,
+            'hourmax' => 24,
+            'comment' => 'emaintenance@expertline.ro',
```
Comment (hupupu) [inline plugin/inc/installsteps/crontasksinstallstep.class.php:59]: This should run during the night also
```suggestion
            'itemtype' => 'PluginIservicePrinterDailyAverageCalculator',
            'name' => 'PrinterDailyAverageCalculator',
            'mode' => CronTask::MODE_EXTERNAL,
            'frequency' => 86400,
            'param' => 0,
            'state' => 1,
            'hourmin' => 4,
            'hourmax' => 7,
            'comment' => 'emaintenance@expertline.ro',
```

File: plugin/inc/printer.class.php [hunk 2]
```diff
@@ -744,9 +744,11 @@ public static function getSerialFieldForEM($table_prefix = ''): string
 
     public static function getCMCondition($cartridge_management_field, $type_field, $state_field): string
     {
+        $blackWhitePrinterType = IserviceToolBox::getIdentifierByAttribute('PrinterType', 'alb-negru');
```
Comment (hupupu) [inline plugin/inc/printer.class.php:747]: I would like to see the `getIdentifierByAttribute` function.

File: plugin/inc/views/printercounters.class.php [hunk 3]
```diff
@@ -314,20 +314,22 @@ protected function getSettings(): array
...
                 left join glpi_suppliers s on s.id = i.suppliers_id
-                left join glpi_plugin_fields_suppliercustomfields cfs on cfs.items_id = s.id and cfs.itemtype = 'Supplier'
+                left join glpi_plugin_fields_suppliersuppliercustomfields cfs on cfs.items_id = s.id and cfs.itemtype = 'Supplier'
                 left join glpi_users u on u.id = p.users_id_tech
                 left join glpi_users ue on ue.id = p.users_id
                 left join glpi_cartridges c on c.printers_id = p.id and c.date_use is not null and c.date_out is null
                 left join glpi_cartridgeitems  ci on ci.id = c.cartridgeitems_id
-                left join glpi_plugin_fields_cartridgeitemcartridgecustomfields cfc on cfc.items_id = ci.id and cfc.itemtype = 'CartridgeItem'
+                left join glpi_plugin_fields_cartridgecartridgecustomfields cfc on cfc.items_id = ci.id and cfc.itemtype = 'CartridgeItem'
+                left join glpi_plugin_fields_cartridgeitemcartridgeitemcustomfields cfci on cfci.items_id = ci.id and cfci.itemtype = 'CartridgeItem'
```
Comment (hupupu) [inline plugin/inc/views/printercounters.class.php:328]: Why did data move from cartridgecustomfield to cartridgeitemcustomfield?

**Review by hupupu — CHANGES_REQUESTED** [review]

File: plugin/inc/printer.class.php (hunk 2 above)
Comment (hupupu) [reply plugin/inc/printer.class.php:747]: Did see, it's ok

File: plugin/inc/views/printercounters.class.php [hunk 4]
```diff
@@ -314,20 +314,22 @@ protected function getSettings(): array
...
                 left join glpi_infocoms i on i.items_id = p.id and i.itemtype = 'Printer'
                 left join glpi_suppliers s on s.id = i.suppliers_id
-                left join glpi_plugin_fields_suppliercustomfields cfs on cfs.items_id = s.id and cfs.itemtype = 'Supplier'
+                left join glpi_plugin_fields_suppliersuppliercustomfields cfs on cfs.items_id = s.id and cfs.itemtype = 'Supplier'
                 left join glpi_users u on u.id = p.users_id_tech
                 left join glpi_users ue on ue.id = p.users_id
                 left join glpi_cartridges c on c.printers_id = p.id and c.date_use is not null and c.date_out is null
                 left join glpi_cartridgeitems  ci on ci.id = c.cartridgeitems_id
-                left join glpi_plugin_fields_cartridgeitemcartridgecustomfields cfc on cfc.items_id = ci.id and cfc.itemtype = 'CartridgeItem'
+                left join glpi_plugin_fields_cartridgecartridgecustomfields cfc on cfc.items_id = c.id and cfc.itemtype = 'Cartridge'
```
Comment (hupupu) [inline plugin/inc/views/printercounters.class.php:327]: Do we need this join?

**Review by hupupu — APPROVED** [review]

File: plugin/inc/views/printers.class.php [hunk 5]
```diff
@@ -321,10 +321,10 @@ protected function getRightCondition($printer_table_alias = 'p.')
 
     protected function getSettings(): array
     {
-        global $CFG_GLPI;
+        global $CFG_GLPI, $CFG_PLUGIN_ISERVICE;
 
         $printer_counters_button = IserviceToolBox::inProfileArray('client') ? '' :
-            "<a class='vsubmit' href='#' target='_blank'>" . __('Printer counters', 'iservice') . " v2</a>";
+            "<a class='vsubmit' href='$CFG_PLUGIN_ISERVICE[root_doc]/front/views.php?view=PrinterCounters' target='_blank'>" . __('Printer counters', 'iservice') . " v2</a>";
```
[the remark this answers is not in the document]
Comment (zsso) [reply plugin/inc/views/printers.class.php:327]: Changed all other occurrences to camel case since that is used everywhere for views.

**Review by hupupu — DISMISSED** [review]

**Pull request merged**